File: apps/server/providers/src/codex/oauth.ts
import * as crypto from "node:crypto";
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import {
  CODEX_ACCOUNT_CLAIM,
  CODEX_AUTHORIZE_URL,
  CODEX_CLIENT_ID,
  CODEX_CALLBACK_PATH,
  CODEX_CALLBACK_PORT,
  CODEX_PROFILE_CLAIM,
  CODEX_SCOPE,
  CODEX_TOKEN_URL,
  CODEX_REFRESH_SKEW_MS,
} from "./constants.js";

export interface CodexOAuthCredential {
  access_token: string;
  refresh_token: string;
  expiresAt: number;
  accountId: string;
  email?: string;
  planType?: string;
}

export interface ParsedCodexCredential {
  accessToken: string;
  refreshToken: string;
  expiresAtMs: number;
  accountId: string;
  email?: string;
}

interface TokenResponse {
  access_token?: string;
  refresh_token?: string;
  id_token?: string;
  expires_in?: number;
}

function credentialPath(): string {
  return process.env.CODEX_CREDENTIALS_PATH ?? path.join(os.homedir(), ".console", "codex-creds.json");
}

export function decodeJwtPayload(token: string): Record<string, unknown> | null {
  try {
    const parts = token.split(".");
    if (parts.length !== 3) return null;
    return JSON.parse(Buffer.from(parts[1]!, "base64url").toString("utf8")) as Record<string, unknown>;
  } catch {
    return null;
  }
}

export function getCodexTokenProfile(
  accessToken: string,
  idToken?: string,
): { accountId?: string; email?: string; planType?: string } {
  const payload = decodeJwtPayload(accessToken);
  const idPayload = idToken ? decodeJwtPayload(idToken) : null;
  const auth = payload?.[CODEX_ACCOUNT_CLAIM] as Record<string, unknown> | undefined;
  const idAuth = idPayload?.[CODEX_ACCOUNT_CLAIM] as Record<string, unknown> | undefined;
  const profile = payload?.[CODEX_PROFILE_CLAIM] as Record<string, unknown> | undefined;
  return {
    accountId: typeof auth?.chatgpt_account_id === "string" ? auth.chatgpt_account_id : undefined,
    email: typeof profile?.email === "string" ? profile.email.trim().toLowerCase() : undefined,
    planType:
      typeof (auth?.chatgpt_plan_type ?? idAuth?.chatgpt_plan_type) === "string"
        ? String(auth?.chatgpt_plan_type ?? idAuth?.chatgpt_plan_type).trim().toLowerCase()
        : undefined,
  };
}

export function generateCodexPkce(): { verifier: string; challenge: string } {
  const verifier = crypto.randomBytes(32).toString("base64url");
  const challenge = new Bun.CryptoHasher("sha256").update(verifier).digest("base64url");
  return { verifier, challenge };
}

export function createCodexAuthorizationUrl(args: {
  state: string;
  verifierChallenge: string;
}): { authUrl: string; redirectUri: string } {
  const redirectUri = `http://localhost:${CODEX_CALLBACK_PORT}${CODEX_CALLBACK_PATH}`;
  const url = new URL(CODEX_AUTHORIZE_URL);
  url.search = new URLSearchParams({
    response_type: "code",
    client_id: CODEX_CLIENT_ID,
    redirect_uri: redirectUri,
    scope: CODEX_SCOPE,
    code_challenge: args.verifierChallenge,
    code_challenge_method: "S256",
    state: args.state,
    id_token_add_organizations: "true",
    codex_cli_simplified_flow: "true",
    originator: "pi",
  }).toString();
  return { authUrl: url.toString(), redirectUri };
}

function tokenError(status: number, body: string): Error {
  let detail = body.trim();
  try {
    const parsed = JSON.parse(detail) as { error?: unknown; error_description?: unknown; message?: unknown };
    detail = String(parsed.error_description ?? parsed.error ?? parsed.message ?? detail);
  } catch {
    // Keep the raw response when it is not JSON.
  }
  return new Error(`Codex OAuth request failed (${status}): ${detail || "unknown error"}`);
}

async function postToken(body: URLSearchParams): Promise<TokenResponse> {
  const response = await fetch(CODEX_TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: body.toString(),
    signal: AbortSignal.timeout(15_000),
  });
  if (!response.ok) throw tokenError(response.status, await response.text().catch(() => ""));
  return (await response.json()) as TokenResponse;
}

export async function exchangeCodexCode(code: string, verifier: string, redirectUri: string): Promise<CodexOAuthCredential> {
  const data = await postToken(
    new URLSearchParams({
      grant_type: "authorization_code",
      client_id: CODEX_CLIENT_ID,
      code,
      code_verifier: verifier,
      redirect_uri: redirectUri,
    }),
  );
  if (!data.access_token || !data.refresh_token || typeof data.expires_in !== "number") {
    throw new Error("Codex OAuth response is missing required token fields.");
  }
  const profile = getCodexTokenProfile(data.access_token, data.id_token);
  if (!profile.accountId) throw new Error("Codex OAuth response did not include a ChatGPT account ID.");
  return {
    access_token: data.access_token,
    refresh_token: data.refresh_token,
    expiresAt: Date.now() + data.expires_in * 1000,
    accountId: profile.accountId,
    email: profile.email,
    planType: profile.planType,
  };
}

async function refreshCodexToken(credential: ParsedCodexCredential): Promise<ParsedCodexCredential> {
  const data = await postToken(
    new URLSearchParams({
      grant_type: "refresh_token",
      client_id: CODEX_CLIENT_ID,
      refresh_token: credential.refreshToken,
    }),
  );
  if (!data.access_token || !data.refresh_token || typeof data.expires_in !== "number") {
    throw new Error("Codex token refresh response is missing required fields.");
  }
  const profile = getCodexTokenProfile(data.access_token);
  const updated: CodexOAuthCredential = {
    access_token: data.access_token,
    refresh_token: data.refresh_token,
    expiresAt: Date.now() + data.expires_in * 1000,
    accountId: profile.accountId ?? credential.accountId,
    email: profile.email ?? credential.email,
  };
  await saveCodexCredential(updated);
  return parseCodexCredential(updated);
}

export function parseCodexCredential(raw: CodexOAuthCredential): ParsedCodexCredential {
  if (!raw.access_token || !raw.refresh_token || !raw.accountId) {
    throw new Error("Invalid Codex credential. Please login again.");
  }
  return {
    accessToken: raw.access_token,
    refreshToken: raw.refresh_token,
    expiresAtMs: raw.expiresAt,
    accountId: raw.accountId,
    email: raw.email,
  };
}

export async function loadCodexCredential(): Promise<ParsedCodexCredential> {
  const envToken = process.env.OPENAI_CODEX_OAUTH_TOKEN;
  if (envToken) {
    const profile = getCodexTokenProfile(envToken);
    if (!profile.accountId) throw new Error("OPENAI_CODEX_OAUTH_TOKEN has no ChatGPT account ID.");
    return {
      accessToken: envToken,
      refreshToken: "",
      expiresAtMs: Number.MAX_SAFE_INTEGER,
      accountId: profile.accountId,
      email: profile.email,
    };
  }
  const raw = JSON.parse(await fs.readFile(credentialPath(), "utf8")) as CodexOAuthCredential;
  return parseCodexCredential(raw);
}

export async function saveCodexCredential(credential: CodexOAuthCredential): Promise<void> {
  const filePath = credentialPath();
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  await fs.writeFile(filePath, JSON.stringify(credential, null, 2), "utf8");
}

export async function codexCredentialExists(): Promise<boolean> {
  if (process.env.OPENAI_CODEX_OAUTH_TOKEN) return true;
  try {
    await fs.access(credentialPath());
    return true;
  } catch {
    return false;
  }
}

export async function refreshCodexIfNeeded(credential: ParsedCodexCredential): Promise<ParsedCodexCredential> {
  if (!credential.refreshToken || Date.now() + CODEX_REFRESH_SKEW_MS < credential.expiresAtMs) return credential;
  return refreshCodexToken(credential);
}

File: apps/server-go/internal/providers/codex/oauth.go
// Codex ChatGPT OAuth: PKCE, authorize URL, code exchange, token refresh,
// credential load/save. Port of apps/server/providers/src/codex/oauth.ts.
// Generic JWT/PKCE/file helpers live in providers/shared.
package codex

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/providers/shared"
)

// OAuthCredential is the stored Codex credential file shape.
type OAuthCredential struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresAt    int64  `json:"expiresAt"`
	AccountID    string `json:"accountId"`
	Email        string `json:"email,omitempty"`
	PlanType     string `json:"planType,omitempty"`
}

// ParsedCredential is the validated in-memory credential.
type ParsedCredential struct {
	AccessToken  string
	RefreshToken string
	ExpiresAtMs  int64
	AccountID    string
	Email        string
}

type tokenResponse struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	IDToken      string `json:"id_token"`
	ExpiresIn    *int64 `json:"expires_in"`
}

// CredentialPath mirrors oauth.ts: CODEX_CREDENTIALS_PATH override,
// otherwise ~/.console/codex-creds.json.
func CredentialPath() string {
	return shared.CredentialPath("CODEX_CREDENTIALS_PATH", "codex-creds.json")
}

// DecodeJWTPayload decodes the middle JWT segment without verification.
func DecodeJWTPayload(token string) map[string]any {
	return shared.DecodeJWTPayload(token)
}

// TokenProfile extracts account id / email / plan from access + id tokens.
func TokenProfile(accessToken, idToken string) (accountID, email, planType string) {
	payload := shared.DecodeJWTPayload(accessToken)
	var idPayload map[string]any
	if idToken != "" {
		idPayload = shared.DecodeJWTPayload(idToken)
	}
	if auth, ok := payload[AccountClaim].(map[string]any); ok {
		if v, ok := auth["chatgpt_account_id"].(string); ok {
			accountID = v
		}
		if v, ok := auth["chatgpt_plan_type"].(string); ok {
			planType = strings.TrimSpace(strings.ToLower(v))
		}
	}
	if planType == "" {
		if auth, ok := idPayload[AccountClaim].(map[string]any); ok {
			if v, ok := auth["chatgpt_plan_type"].(string); ok {
				planType = strings.TrimSpace(strings.ToLower(v))
			}
		}
	}
	if profile, ok := payload[ProfileClaim].(map[string]any); ok {
		if v, ok := profile["email"].(string); ok {
			email = strings.TrimSpace(strings.ToLower(v))
		}
	}
	return accountID, email, planType
}

// GeneratePKCE creates a verifier/challenge pair (S256).
func GeneratePKCE() (verifier, challenge string, err error) {
	return shared.GeneratePKCE()
}

// AuthorizationURL builds the ChatGPT OAuth authorize URL + redirect URI.
func AuthorizationURL(state, verifierChallenge string) (authURL, redirectURI string) {
	redirectURI = fmt.Sprintf("http://localhost:%d%s", CallbackPort, CallbackPath)
	u, _ := url.Parse(AuthorizeURL)
	q := url.Values{
		"response_type":              {"code"},
		"client_id":                  {ClientID},
		"redirect_uri":               {redirectURI},
		"scope":                      {Scope},
		"code_challenge":             {verifierChallenge},
		"code_challenge_method":      {"S256"},
		"state":                      {state},
		"id_token_add_organizations": {"true"},
		"codex_cli_simplified_flow":  {"true"},
		"originator":                 {"pi"},
	}
	u.RawQuery = q.Encode()
	return u.String(), redirectURI
}

func postToken(client *http.Client, body url.Values) (tokenResponse, error) {
	if client == nil {
		client = &http.Client{Timeout: 15 * time.Second}
	}
	req, err := http.NewRequest("POST", TokenURL, strings.NewReader(body.Encode()))
	if err != nil {
		return tokenResponse{}, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	resp, err := client.Do(req)
	if err != nil {
		return tokenResponse{}, err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return tokenResponse{}, shared.TokenError("Codex", resp.StatusCode, string(raw))
	}
	var data tokenResponse
	if err := json.Unmarshal(raw, &data); err != nil {
		return tokenResponse{}, fmt.Errorf("decode Codex token response: %w", err)
	}
	return data, nil
}

// ExchangeCode exchanges an authorization code for stored credentials.
func ExchangeCode(client *http.Client, code, verifier, redirectURI string) (OAuthCredential, error) {
	data, err := postToken(client, url.Values{
		"grant_type":    {"authorization_code"},
		"client_id":     {ClientID},
		"code":          {code},
		"code_verifier": {verifier},
		"redirect_uri":  {redirectURI},
	})
	if err != nil {
		return OAuthCredential{}, err
	}
	if data.AccessToken == "" || data.RefreshToken == "" || data.ExpiresIn == nil {
		return OAuthCredential{}, fmt.Errorf("Codex OAuth response is missing required token fields.")
	}
	accountID, email, planType := TokenProfile(data.AccessToken, data.IDToken)
	if accountID == "" {
		return OAuthCredential{}, fmt.Errorf("Codex OAuth response did not include a ChatGPT account ID.")
	}
	return OAuthCredential{
		AccessToken:  data.AccessToken,
		RefreshToken: data.RefreshToken,
		ExpiresAt:    time.Now().UnixMilli() + *data.ExpiresIn*1000,
		AccountID:    accountID,
		Email:        email,
		PlanType:     planType,
	}, nil
}

func refreshToken(client *http.Client, cred ParsedCredential) (ParsedCredential, error) {
	data, err := postToken(client, url.Values{
		"grant_type":    {"refresh_token"},
		"client_id":     {ClientID},
		"refresh_token": {cred.RefreshToken},
	})
	if err != nil {
		return ParsedCredential{}, err
	}
	if data.AccessToken == "" || data.RefreshToken == "" || data.ExpiresIn == nil {
		return ParsedCredential{}, fmt.Errorf("Codex token refresh response is missing required fields.")
	}
	accountID, email, _ := TokenProfile(data.AccessToken, "")
	if accountID == "" {
		accountID = cred.AccountID
	}
	if email == "" {
		email = cred.Email
	}
	updated := OAuthCredential{
		AccessToken:  data.AccessToken,
		RefreshToken: data.RefreshToken,
		ExpiresAt:    time.Now().UnixMilli() + *data.ExpiresIn*1000,
		AccountID:    accountID,
		Email:        email,
	}
	if err := SaveCredential(updated); err != nil {
		return ParsedCredential{}, err
	}
	return ParseCredential(updated)
}

// ParseCredential validates a stored credential.
func ParseCredential(raw OAuthCredential) (ParsedCredential, error) {
	if raw.AccessToken == "" || raw.RefreshToken == "" || raw.AccountID == "" {
		return ParsedCredential{}, fmt.Errorf("Invalid Codex credential. Please login again.")
	}
	return ParsedCredential{
		AccessToken:  raw.AccessToken,
		RefreshToken: raw.RefreshToken,
		ExpiresAtMs:  raw.ExpiresAt,
		AccountID:    raw.AccountID,
		Email:        raw.Email,
	}, nil
}

// LoadCredential reads OPENAI_CODEX_OAUTH_TOKEN or the credential file.
func LoadCredential() (ParsedCredential, error) {
	if envToken := os.Getenv("OPENAI_CODEX_OAUTH_TOKEN"); envToken != "" {
		accountID, email, _ := TokenProfile(envToken, "")
		if accountID == "" {
			return ParsedCredential{}, fmt.Errorf("OPENAI_CODEX_OAUTH_TOKEN has no ChatGPT account ID.")
		}
		return ParsedCredential{
			AccessToken:  envToken,
			RefreshToken: "",
			ExpiresAtMs:  1<<62 - 1,
			AccountID:    accountID,
			Email:        email,
		}, nil
	}
	raw, err := os.ReadFile(CredentialPath())
	if err != nil {
		return ParsedCredential{}, err
	}
	// Tolerate extra whitespace/BOM-free JSON only (parity with TS readFile).
	dec := json.NewDecoder(bytes.NewReader(bytes.TrimSpace(raw)))
	var cred OAuthCredential
	if err := dec.Decode(&cred); err != nil {
		return ParsedCredential{}, err
	}
	return ParseCredential(cred)
}

// SaveCredential writes the credential file (0600 dir 0700).
func SaveCredential(cred OAuthCredential) error {
	return shared.SaveCredentialFile(CredentialPath(), cred)
}

// CredentialExists reports whether a login exists (env token or file).
func CredentialExists() bool {
	if os.Getenv("OPENAI_CODEX_OAUTH_TOKEN") != "" {
		return true
	}
	return shared.CredentialFileExists(CredentialPath())
}

// RefreshIfNeeded refreshes an expiring credential (60s skew).
func RefreshIfNeeded(client *http.Client, cred ParsedCredential) (ParsedCredential, error) {
	if cred.RefreshToken == "" || time.Now().UnixMilli()+RefreshSkewMs < cred.ExpiresAtMs {
		return cred, nil
	}
	return refreshToken(client, cred)
}
